crates/modelardb_server/src/storage/types.rs

```diff
@@ -0,0 +1,291 @@
+            ),
+            remaining_compressed_memory_in_bytes: RwLock::new(
+                compressed_memory_in_bytes.try_into().unwrap(),
+            ),
+        }
+    }
+
+    /// Change the amount of memory available for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata by `size_in_bytes`.
+    pub(super) fn update_uncompressed_memory(&self, size_in_bytes: isize) {
+        // unwrap() is safe as write() only returns an error if the lock is poisoned.
+        *self.remaining_uncompressed_memory_in_bytes.write().unwrap() += size_in_bytes
+    }
+
+    /// Return the amount of memory available for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata in bytes.
+    pub(super) fn remaining_uncompressed_memory_in_bytes(&self) -> isize {
+        // unwrap() is safe as write() only returns an error if the lock is poisoned.
+        *self.remaining_uncompressed_memory_in_bytes.read().unwrap()
+    }
+
+    /// Try to reserve `size_in_bytes` bytes of memory for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata. Returns [`true`] if the reservation
+    /// succeeds and [`false`] otherwise.
+    pub(super) fn try_reserve_uncompressed_memory(&self, size_in_bytes: usize) -> bool {
+        // unwrap() is safe as write() only returns an error if the lock is poisoned.
+        let mut remaining_uncompressed_memory_in_bytes =
+            self.remaining_uncompressed_memory_in_bytes.write().unwrap();
+
+        let size_in_bytes = size_in_bytes as isize;
+
+        if size_in_bytes <= *remaining_uncompressed_memory_in_bytes {
+            *remaining_uncompressed_memory_in_bytes -= size_in_bytes;
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Free `size_in_bytes` bytes of memory for storing
+    /// [`RecordBatches`](datafusion::arrow::record_batch::RecordBatch) containing multivariate time
+    /// series with metadata and
+    /// [`UncompressedDataBuffers`](crate::storage::uncompressed_data_buffer::UncompressedDataBuffer)
+    /// containing univariate time series without metadata. Returns [`true`] if the reservation
+    /// succeeds and [`false`] otherwise.
+    pub(super) fn free_uncompressed_memory(&self, size_in_bytes: usize) {
```

Comment:
It can be kept for convenience but since this method is very similar to `update_uncompressed_memory`, an argument could be made for removing this and only using `update_uncompressed_memory`.  

If this is removed, also remove `free_compressed_memory`.

Reply:
I also contemplated removing the `free_` methods and using the `update_`/`adjust_` methods for both freeing and adjusting memory, but I kept the `free_` methods as they can take an `usize` as input (freeing a negative amount of memory is not a thing) and thus prevents accidental freeing a negative amount of memory and the amount of code duplication is very small. However, the trade-off should definitely be reevaluated if the methods are extended.